### DISCUSSION TITLE
My WASM app do not load (anymore) on Android's Chrome (crash in StandardAssetLoader ctor)

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#16738] [Q&A]
Hi,
I have the following exception (using remote debug) on Android's Chrome and my app is stuck on the splashscreen.
![image](https://github.com/user-attachments/assets/7e7276f9-1e8a-4a83-a2df-72005499297a)
**Could someone explain what StandardAssetLoader is used for?** I don't think I am using it myself, it should be internal to Avalonia, and I don't know how to debug.

I have no problem on desktop PC (chrome or firefox), **and the website load normally on my android device if I use Edge instead of Chrome, so may be the problem is not in my code? but a bug in Avalonia?** 

I do not know since when my app started to crash on my smartphone.

**Comment by workgroupengineering** [comment]
most likely the line that generates the exception is this:

https://github.com/AvaloniaUI/Avalonia/blob/dd17e85271726135ea5e87b6b28f4d371c2f9a40/src/Avalonia.Base/Platform/Internal/AssemblyDescriptor.cs#L35-L36

perhaps the same asset has been uploaded multiple times. Check your `Project.Browser.csproj`

maybe @MrJul  might know more.

**Comment by dbriard** [comment]
Thanks for your reply @workgroupengineering 

I tried to debug by adding the code of AssemblyDescriptor in my program.cs and with GetExecutingAssembly() there is only 1 resources : 
![image](https://github.com/user-attachments/assets/ef7d3ced-e654-49ab-91c8-06d5a7a47df1)

No sure what assembly is used in AssemblyDescriptor when I got the error... 

If it this is really a duplicated resource (I don't know how I could do that if I would), I should have the issue on all browsers I think.

I will try to install a simulator, or to restore an old version of my app.

**Reply by workgroupengineering** [reply]
> Thanks for your reply @workgroupengineering
> 
> I tried to debug by adding the code of AssemblyDescriptor in my program.cs and with GetExecutingAssembly() there is only 1 resources : ![image](https://private-user-images.githubusercontent.com/3595443/359463292-ef7d3ced-e654-49ab-91c8-06d5a7a47df1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNTkxMDksIm5iZiI6MTcyNDE1ODgwOSwicGF0aCI6Ii8zNTk1NDQzLzM1OTQ2MzI5Mi1lZjdkM2NlZC1lNjU0LTQ5YWItOTFjOC0wNmQ1YTdhNDdkZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjBUMTMwMDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTM5NGJiOGJiNTIyMTczYmUxNjIyOWNhODM5ZjJlYzFkYmQxZDY0YjIwYWE3ZTk0MjA3NmNmNjk5Njk1ZDk0ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7KVvxThzgSyFq7YS0Lwj790c0CPXmNeaRJAapoXVQuI)
> 
> No sure what assembly is used in AssemblyDescriptor when I got the error...
> 
> If it this is really a duplicated resource (I don't know how I could do that if I would), I should have the issue on all browsers I think.
> 
> I will try to install a simulator, or to restore an old version of my app.

the duplicate key should be inside `!AvaloniaResources`, use code similar to the following to enumerate the keys:

```c#
        using var resources = assembly.GetManifestResourceStream("!AvaloniaResources");
        if (resources != null)
        {
            var indexLength = new BinaryReader(resources).ReadInt32();
            var index = AvaloniaResourcesIndexReaderWriter.ReadIndex(new SlicedStream(resources, 4, indexLength));
            
           foreach (var item in index)
           {
                Console.WriteLine($"Path: {GetPathRooted(Item)} Offset: {item.Offset} Size: {item.Size}");
           }
       }

      static string GetPathRooted(AvaloniaResourcesIndexEntry r) =>
       r.Path![0] == '/' ? r.Path : '/' + r.Path;
```

**Reply by dbriard** [reply]
So far I tried:

- Run the website in Chrome (old version) in an old android simulator => WORK
- Run hello world app on my phone => WORK
- Run a 3-month-ago version of my app on my phone => Same EXCEPTION (I am almost sure it worked 3 month ago)
- I wanted to try a simulator with latest version of Chrome but for any reason, I get a black screen on all websites in that simulator... trying to found why.

I will try on my wife's phone later... may be something wrong with current Chrome version (I tried Chrome beta and same error), or with my phone :( I hate those mystery bugs.

I also can check all assemblies for duplicated resources... 

Edit: @workgroupengineering thanks for the code repro, I will try!

**Reply by dbriard** [reply]
So,
On Chrome Desktop:
![image](https://github.com/user-attachments/assets/0c5f5b45-83b8-4721-af14-d55657e275a9)

Chrome Android Simulator (old version of Chrome)
![image](https://github.com/user-attachments/assets/da8c8a88-cc87-44ad-b9a5-dbbbe5011f35)
![image](https://github.com/user-attachments/assets/4e9b6612-0a63-4ea9-90f2-56b59ccefa4c)

Chrome Android Simulator (latest version of Chrome)
![image](https://github.com/user-attachments/assets/85070e6c-c22b-4d78-9214-2922552efc43)
![image](https://github.com/user-attachments/assets/073aa92a-3b9c-4579-9a29-9d1c36290c7e)

But on Chrome Android on my device Nokia 7.1 (Chrome latest version 127.0.6533.103): 
![image](https://github.com/user-attachments/assets/07df8609-db7b-4d4e-a0c1-d45ea501946b)

WTF?!!! Could it be a bug related to real hardware? a recent change in Avalonia? or Chrome android? maybe .NET

Both resource name are missing the first 15 characters...

**Reply by workgroupengineering** [reply]
the path is read here.

https://github.com/AvaloniaUI/Avalonia/blob/2d516027602698b1aa817f46edc62fd9d337c66a/src/Avalonia.Base/Utilities/AvaloniaResourcesIndex.cs#L41C35-L41C45

since the offset and size match. I think there is some problem in the implementation of  `BinaryReader.ReadString()` in wasm.

**Reply by dbriard** [reply]
> the path is read here.
> 
> https://github.com/AvaloniaUI/Avalonia/blob/2d516027602698b1aa817f46edc62fd9d337c66a/src/Avalonia.Base/Utilities/AvaloniaResourcesIndex.cs#L41C35-L41C45
> 
> since the offset and size match. I think there is some problem in the implementation of `BinaryReader.ReadString()` in wasm.

Not on all WASM as it work on Edge on the same device. I believed that both Edge and Chrome was based on the same engine, but may be not on Android.

Will try to display the entire stream in hex format, and compare tomorrow. That bug is really annoying... 

Thx again for your help.

**Reply by dbriard** [reply]
Same error with a different Android phone and Chrome.

**Reply by dbriard** [reply]
I compared the content of the resource stream by displaying as hex string, and it is the same on my computer than on my android phone.
I retried a few past builds, and it seems the problem has started when I migrated my project to Avaloinia 11.1 and .NET8.

**Reply by workgroupengineering** [reply]
Try using `BitConverter.IsLittleEndian` to check byte order

**Reply by dbriard** [reply]
I copied the code of ReadIndex directly in my program.cs and also copied the code of BinaryReader ReadString in my code with a few WriteLines.

On Desktop (or Android Edge, or Android simulator)
![image](https://github.com/user-attachments/assets/838b7987-ae03-4379-845a-f78e6ad32699)

On Android Chrome read device
![image](https://github.com/user-attachments/assets/1a4def92-167f-4eea-8e25-51aa5a859c45)

It look like the issue in either in _decoder.GetChars (=Encoding.UTF8.GetDecoder()) or in StringBuilder.Append
![image](https://github.com/user-attachments/assets/013c30f1-e057-48df-b2b1-41c955760e89)

But that do not explain why I got a duplicated key exception... as even if I get weird names, they are differents.

**Reply by dbriard** [reply]
Bug is in Encoding.UTF8.GetDecoder().GetChars()

GetChars() on the same byte array do not give the same result, so it is most likely a bug in .NET :(

Desktop:
![image](https://github.com/user-attachments/assets/7bda8c5e-d746-4161-8857-f84520296a31)

Android device Chrome:
![image](https://github.com/user-attachments/assets/3bbe9952-2ce9-498b-9ecb-e14258b920bd)

But unfortunatly I cannot debug deeper:

UTF8Encoding class:
![image](https://github.com/user-attachments/assets/a8744299-7c3a-4850-b372-d1c563e730ab)

**Reply by MrJul** [reply]
Since you don't reproduce with Edge, it looks like a bug in Chrome on ARM rather than a bug in .NET (even though both browsers are based on Chromium, they have differences), especially if your old app that used to work now doesn't. Can you also try Firefox?

**Reply by dbriard** [reply]
@MrJul @workgroupengineering 
No problem using Firefox on Android.
I tried to compile without AOT, and it works on Chrome Android!!!

So yes, may be a bug in Chrome Android, but the same device with the same Chrome work with old build (.NET 7 Avalonia 11.0).
I don't thing but may be the code that read AvaloniaResources has changed in 11.1 and that may explain why it was working before...
And I don't know if the BinaryReader.ReadString() is used a lot in Avalonia, but if it is used only once here, may be that can be replaced  by a custom implementation of ReadString and solve my issue... 

At least one good news: I just tested on iPhone (Chrome and Safari) and this is the opposite... my Avalonia 11.0/.NET7 never worked on iPhone, and now it work! with AOT and without AOT, with Chrome and with Safari...

So the bad combination is Android + ARM + Chrome + AOT + .NET8 !

**Reply by dbriard** [reply]
I removed my avalonia resource in Browser project, but still get the duplicated key error.. 
I suppose that this is because of avalonia resources in other assemblies that have a short name and are resolved by ReadString as string.empty... 

I cannot disable AOT as performance is key factor of that app.

That would be great of we can provide our own class to replace StandardAssetLoader, especially as this class is flagged as Unstable in the avalonia code.
